kernel/libc/koslib/gmon.c

```diff
@@ -0,0 +1,272 @@
+                cxt->samples[index]++;
+            }
+            else {
+                printf("Outside text range pc(histogram_callback): %p\n", (void *)pc);
+            }
+
+            thd_sleep(HISTOGRAM_INTERVAL_MS);
+        }
+
+    }
+
+    return NULL;
+}
+
+void moncontrol(bool mode) {
+    gmon_context_t *cxt = &g_context;
+
+    /* Don't change the state if we ran into an error */
+    if(cxt->state == GMON_PROF_ERROR)
+        return;
+
+    /* Treat start request as stop if error or gmon not initialized */
+    if(mode && cxt->initialized) {
+        /* Start */
+        cxt->state = GMON_PROF_ON;
+    } else {
+        /* Stop */
+        cxt->state = GMON_PROF_OFF;
+    }
+}
+
+void _mcount(uintptr_t frompc, uintptr_t selfpc) {
+    size_t index;
+    gmon_context_t *cxt = &g_context;
+
+    if(cxt->state != GMON_PROF_ON)
+        return;
+
+    /* Make sure call is from within text section */
+    if(frompc >= cxt->lowpc && frompc <= cxt->highpc) {
+        cxt->pc = selfpc; /* Save the last place we are */
+        index = (frompc - cxt->lowpc) / cxt->hashfraction;
+        cxt->arcs[index].frompc = frompc;
+        cxt->arcs[index].selfpc = selfpc;
+        cxt->arcs[index].count++;
+    }
+}
+
+void __monstartup(uintptr_t lowpc, uintptr_t highpc) {
+    gmon_context_t *cxt = &g_context;
+
+    /* Exit early if we already initialized */
+    if(cxt->initialized)
+        return;
+
+    cxt->initialized = true;
+
+    cxt->lowpc = ROUNDDOWN(lowpc, HISTFRACTION);
+    cxt->highpc = ROUNDUP(highpc, HISTFRACTION);
+    cxt->textsize = cxt->highpc - cxt->lowpc;
+    cxt->hashfraction = HISTFRACTION;
+
+    cxt->narcs = (cxt->textsize + cxt->hashfraction - 1) / cxt->hashfraction;
+    cxt->arcs = (gmon_arc_t *)calloc(cxt->narcs, sizeof(gmon_arc_t));
+    if(!cxt->arcs) {
+        cxt->state = GMON_PROF_ERROR;
+        return;
+    }
+
+    cxt->nsamples = (cxt->textsize + cxt->hashfraction - 1) / cxt->hashfraction;
+    cxt->samples = (uint32_t *)calloc(cxt->nsamples, sizeof(uint32_t));
+    if(!cxt->samples) {
+        free(cxt->arcs);
+        cxt->arcs = NULL;
+        cxt->state = GMON_PROF_ERROR;
+        return;
+    }
+
+    printf("Low: %p, High: %p\n", (void *)cxt->lowpc, (void *)cxt->highpc);
+    printf("Textsize: %d\n", cxt->textsize);
+    printf("narcs: %d\n", cxt->narcs);
+    printf("nsamples: %d\n", cxt->nsamples);
+    printf("[Bytes allocated] Arcs: %d bytes, Histogram: %d bytes, Total: %d\n",
+        sizeof(gmon_arc_t) * cxt->narcs,
+        cxt->nsamples * sizeof(uint32_t),
+        sizeof(gmon_arc_t) * cxt->narcs + cxt->nsamples * sizeof(uint32_t));
+
+    /* Initialize histogram thread related members */
+    cxt->running_thread = true;
+    cxt->histogram_thread = thd_create(false, histogram_callback, NULL);
+    thd_set_prio(cxt->histogram_thread, PRIO_DEFAULT / 2);
+
+    /* Turn profiling on */
+    moncontrol(true);
+}
+
+void _mcleanup(void) {
```

Comment:
I think this should probably just be a single underscore. I did see GCC uses 2 for that symbol (https://github.com/gcc-mirror/gcc/blob/master/libgcc/config/sh/crt1.S#L184) but not sure there's a specific need to match since we use our own startup asm, but perhaps I'm missing something?

Reply:
No specific need.  Just wanted to match.  I'll make it use one.